pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,234 @@
+package ingest
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/multierror"
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+)
+
+// consumerGroup is only used to store commit offsets, not for actual consuming.
+const consumerGroup = "mimir"
+
+type Record struct {
+	TenantID string
+	Content  []byte
+}
+
+type RecordConsumer interface {
+	Consume(context.Context, Record) error
+}
+
+type PartitionReader struct {
+	services.Service
+
+	kafkaAddress string
+	kafkaTopic   string
+	partition    int32
+
+	client *kgo.Client
+
+	consumer RecordConsumer
+	metrics  *readerMetrics
+
+	logger log.Logger
+	reg    prometheus.Registerer
+}
+
+func NewReader(kafkaAddress, kafkaTopic string, partitionID int32, consumer RecordConsumer, logger log.Logger, reg prometheus.Registerer) (*PartitionReader, error) {
+	metrics := newReaderMetrics(reg)
+	r := &PartitionReader{
+		kafkaAddress: kafkaAddress,
+		kafkaTopic:   kafkaTopic,
+		partition:    partitionID,
+		reg:          reg,
+		consumer:     consumer, // TODO consume records in parallel
+		metrics:      metrics,
+		logger:       log.With(logger, "partition", partitionID),
+	}
+
+	r.Service = services.NewBasicService(r.start, r.run, r.stop)
+	return r, nil
+}
+
+func (r *PartitionReader) start(ctx context.Context) error {
+	offset, err := r.fetchLastCommittedOffset(ctx)
```

Comment:
This is something we should retry (with backoff) for a while before giving up. If there's a transient error, I don't want to restart the whole ingester when it could be enough to retry it.